test/asynchronous/test_database.py

```diff
@@ -0,0 +1,778 @@
+        write_concern = WriteConcern(w=2, j=True)
+        read_concern = ReadConcern("majority")
+        coll = self.client.pymongo_test.get_collection(
+            "foo", codec_options, ReadPreference.SECONDARY, write_concern, read_concern
+        )
+        self.assertEqual("foo", coll.name)
+        self.assertEqual(codec_options, coll.codec_options)
+        self.assertEqual(ReadPreference.SECONDARY, coll.read_preference)
+        self.assertEqual(write_concern, coll.write_concern)
+        self.assertEqual(read_concern, coll.read_concern)
+
+    def test_getattr(self):
+        db = self.client.pymongo_test
+        self.assertTrue(isinstance(db["_does_not_exist"], AsyncCollection))
+
+        with self.assertRaises(AttributeError) as context:
+            db._does_not_exist
+
+        # Message should be: "AttributeError: Database has no attribute
+        # '_does_not_exist'. To access the _does_not_exist collection,
+        # use database['_does_not_exist']".
+        self.assertIn("has no attribute '_does_not_exist'", str(context.exception))
+
+    def test_iteration(self):
+        db = self.client.pymongo_test
+        if "PyPy" in sys.version and sys.version_info < (3, 8, 15):
+            msg = "'NoneType' object is not callable"
+        else:
+            if _IS_SYNC:
+                msg = "'Database' object is not iterable"
+            else:
+                msg = "'AsyncDatabase' object is not iterable"
+        # Iteration fails
+        with self.assertRaisesRegex(TypeError, msg):
+            for _ in db:  # type: ignore[misc] # error: "None" not callable  [misc]
+                break
+        # Index fails
+        with self.assertRaises(TypeError):
+            _ = db[0]
+        # next fails
+        with self.assertRaisesRegex(TypeError, "'Database' object is not iterable"):
+            _ = next(db)
+        # .next() fails
+        with self.assertRaisesRegex(TypeError, "'Database' object is not iterable"):
+            _ = db.next()
+        # Do not implement typing.Iterable.
+        self.assertNotIsInstance(db, Iterable)
+
+
+class TestDatabase(AsyncIntegrationTest):
+    def test_equality(self):
+        self.assertNotEqual(AsyncDatabase(self.client, "test"), AsyncDatabase(self.client, "mike"))
+        self.assertEqual(AsyncDatabase(self.client, "test"), AsyncDatabase(self.client, "test"))
+
+        # Explicitly test inequality
+        self.assertFalse(AsyncDatabase(self.client, "test") != AsyncDatabase(self.client, "test"))
+
+    def test_hashable(self):
+        self.assertIn(self.client.test, {AsyncDatabase(self.client, "test")})
+
+    def test_get_coll(self):
+        db = AsyncDatabase(self.client, "pymongo_test")
+        self.assertEqual(db.test, db["test"])
+        self.assertEqual(db.test, AsyncCollection(db, "test"))
+        self.assertNotEqual(db.test, AsyncCollection(db, "mike"))
+        self.assertEqual(db.test.mike, db["test.mike"])
+
+    def test_repr(self):
+        if _IS_SYNC:
```

Comment:
The synchronization process will take care of the renaming for us, this check is unnecessary.